libbeat/cmd/export/template.go

```diff
@@ -21,84 +21,41 @@ import (
 	"fmt"
 	"os"
 
+	"github.com/elastic/beats/libbeat/idxmgmt/ilm"
+
 	"github.com/spf13/cobra"
 
+	"github.com/elastic/beats/libbeat/beat"
 	"github.com/elastic/beats/libbeat/cmd/instance"
 	"github.com/elastic/beats/libbeat/common"
 	"github.com/elastic/beats/libbeat/idxmgmt"
 	"github.com/elastic/beats/libbeat/logp"
-	"github.com/elastic/beats/libbeat/paths"
-	"github.com/elastic/beats/libbeat/template"
 )
 
 func GenTemplateConfigCmd(settings instance.Settings) *cobra.Command {
 	genTemplateConfigCmd := &cobra.Command{
 		Use:   "template",
 		Short: "Export index template to stdout",
 		Run: func(cmd *cobra.Command, args []string) {
-			version, _ := cmd.Flags().GetString("es.version")
-			index, _ := cmd.Flags().GetString("index")
 			noILM, _ := cmd.Flags().GetBool("noilm")
-
-			b, err := instance.NewBeat(settings.Name, settings.IndexPrefix, settings.Version)
-			if err != nil {
-				fatalf("Error initializing beat: %+v", err)
-			}
-			err = b.InitWithSettings(settings)
-			if err != nil {
-				fatalf("Error initializing beat: %+v", err)
-			}
-
-			if version == "" {
-				version = b.Info.Version
-			}
-			esVersion, err := common.NewVersion(version)
-			if err != nil {
-				fatalf("Invalid Elasticsearch version: %+v", err)
-			}
-
-			imFactory := settings.IndexManagement
-			if imFactory == nil {
-				imFactory = idxmgmt.MakeDefaultSupport(settings.ILM)
-			}
-			indexManager, err := imFactory(logp.NewLogger("index-management"), b.Info, b.RawConfig)
-			if err != nil {
-				fatalf("Error initializing the index manager: %+v", err)
+			if noILM {
+				settings.ILM = ilmNoopSupport
 			}
 
-			tmplCfg, err := indexManager.TemplateConfig(!noILM)
+			b, err := instance.NewInitializedBeat(settings)
 			if err != nil {
-				fatalf("Template error detected: %+v", err)
-			}
-			if tmplCfg.Enabled == false {
-				tmplCfg = template.DefaultConfig()
-			}
-
-			tmpl, err := template.New(b.Info.Version, index, *esVersion, tmplCfg, b.Config.Migration.Enabled())
-			if err != nil {
-				fatalf("Error generating template: %+v", err)
+				fatalf("error initializing beat: %+v", err)
 			}
 
-			var templateString common.MapStr
-			if tmplCfg.Fields != "" {
-				fieldsPath := paths.Resolve(paths.Config, tmplCfg.Fields)
-				templateString, err = tmpl.LoadFile(fieldsPath)
-			} else {
-				templateString, err = tmpl.LoadBytes(b.Fields)
-			}
-			if err != nil {
-				fatalf("Error generating template: %+v", err)
-			}
-
-			_, err = os.Stdout.WriteString(templateString.StringToPrint() + "\n")
-			if err != nil {
-				fatalf("Error writing template: %+v", err)
+			idxManager := b.IdxMgmtSupporter().Manager(nil, idxmgmt.BeatsAssets(b.Fields))
```

Comment:
passing `nil` for the `client` paramter will force the manager to write to stdout?

Reply:
Yes. If there is no client passed in, stdout will be the default handling. 
It would be nicer to pass in a more generic handler upfront, that can encapsulate a stdout, file, es client etc. But this would require us to change from passing in a `ESClient` implementation to a more generic construct and, again, I didn't want to make too many changes at once. My thoughts were, that we might introduce a `file` handler additionally to the `stdout` handler in the future, to print policies and templates each to a file. I think that would be a perfect option to also rework the current `ESClient` interface then.